File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/hash/Crc32c.java
package com.freight.hash;

/** crc32c over raw bytes. */
public final class Crc32c implements HashAlgorithm {

    @Override
    public String name() {
        return "crc32c";
    }

    @Override
    public long apply(byte[] data) {
        long crc = 0xFFFFFFFFL;
        for (int i = 0; i < data.length; i++) {
            crc ^= (data[i] & 0xFF);
            for (int bit = 0; bit < 8; bit++) {
                crc = ((crc & 1L) != 0L) ? ((crc >>> 1) ^ 0x82F63B78L) : (crc >>> 1);
            }
        }
        return (crc ^ 0xFFFFFFFFL) & 0xFFFFFFFFL;
    }
}

File: freight-triage-polyglot-ledger-recovery/environment/native/src/hash/crc32c.cpp
#include "freight/hashes.h"

namespace freight {

// crc32c over raw bytes.
uint64_t hash_crc32c(const std::string& data) {
  uint32_t crc = 0xFFFFFFFFu;
  for (size_t i = 0; i < data.size(); ++i) {
    crc ^= static_cast<uint8_t>(data[i]);
    for (int bit = 0; bit < 8; ++bit) {
      crc = (crc & 1u) ? ((crc >> 1) ^ 0x82F63B78u) : (crc >> 1);
    }
  }
  return static_cast<uint64_t>(crc ^ 0xFFFFFFFFu);
}

}  // namespace freight
